### PR TITLE
Extract distributions/discrete/UniformInt from UniformIntVertex.

### PR DESCRIPTION
**Pull request by icantsayrural**
Will use it for computing KL divergence.

Also added a fix for #197.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/UniformInt.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public DoubleTensor logProb(IntegerTensor x) {
+        DoubleTensor maxBound = xMax.toDouble();
+        DoubleTensor minBound = xMin.toDouble();
+        DoubleTensor xDouble = x.toDouble();
+
+        DoubleTensor logOfWithinBounds = maxBound.minus(minBound).logInPlace().unaryMinusInPlace();
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(xDouble.getGreaterThanMask(maxBound), Double.NEGATIVE_INFINITY);
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(xDouble.getLessThanOrEqualToMask(minBound), Double.NEGATIVE_INFINITY);
+
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/UniformInt.java]: Might also be worth adding a test for the prob calculation as well?  It doesn't look like this is tested in UniformIntVertexTest at the moment?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/UniformInt.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public DoubleTensor logProb(IntegerTensor x) {
+        DoubleTensor maxBound = xMax.toDouble();
+        DoubleTensor minBound = xMin.toDouble();
+        DoubleTensor xDouble = x.toDouble();
+
+        DoubleTensor logOfWithinBounds = maxBound.minus(minBound).logInPlace().unaryMinusInPlace();
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(xDouble.getGreaterThanMask(maxBound), Double.NEGATIVE_INFINITY);
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(xDouble.getLessThanOrEqualToMask(minBound), Double.NEGATIVE_INFINITY);
```
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/UniformInt.java]: Yeah, nice catch. I see the same bug in continuous uniform as well.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/UniformVertexTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        UniformVertex testUniformVertex = new UniformVertex(new int[]{1, N}, lowerBound, upperBound);
+        assertEquals(testUniformVertex.logProb(Nd4jDoubleTensor.scalar(upperBound)), Double.NEGATIVE_INFINITY, 1e-6);
+    }
+
+    @Test
+    public void logProbLowerBoundIsNotNegativeInfinity() {
+        UniformVertex testUniformVertex = new UniformVertex(new int[]{1, N}, lowerBound, upperBound);
+        assertNotEquals(testUniformVertex.logProb(Nd4jDoubleTensor.scalar(lowerBound)), Double.NEGATIVE_INFINITY, 1e-6);
+    }
+}
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/UniformVertexTest.java]: Do you think it's worth adding a test like "sampleMatchesLogProb" from the other vertices to the test set?  Obviously this is a very simple distribution, but I think the work to do to get it added is pretty minimal too.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Uniform.java [hunk 4]
```diff
@@ -38,8 +38,8 @@ public DoubleTensor sample(int[] shape, KeanuRandom random) {
     public DoubleTensor logProb(DoubleTensor x) {
 
         DoubleTensor logOfWithinBounds = xMax.minus(xMin).logInPlace().unaryMinusInPlace();
-        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(x.getGreaterThanMask(xMax), Double.NEGATIVE_INFINITY);
-        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(x.getLessThanOrEqualToMask(xMin), Double.NEGATIVE_INFINITY);
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(x.getGreaterThanOrEqualToMask(xMax), Double.NEGATIVE_INFINITY);
+        logOfWithinBounds = logOfWithinBounds.setWithMaskInPlace(x.getLessThanMask(xMin), Double.NEGATIVE_INFINITY);
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/Uniform.java]: Awesome - Thanks for fixing this and spotting the same problem was in Uniform too.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/UniformIntVertexTest.java [hunk 5]
```diff
@@ -42,7 +43,7 @@ public void allSamplesAreWithinBounds() {
         Integer maxSample = Collections.max(samples);
 
         assertTrue(minSample >= lowerBound);
-        assertTrue(maxSample <= upperBound);
+        assertTrue(maxSample < upperBound);
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/UniformIntVertexTest.java]: Good change.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/UniformIntVertexTest.java [hunk 6]
```diff
@@ -73,4 +74,16 @@ public void canUseFullIntRange() {
...
+    public void logProbUpperBoundIsNegativeInfinity() {
+        UniformIntVertex testUniformVertex = new UniformIntVertex(new int[]{1, N}, lowerBound, upperBound);
+        assertEquals(testUniformVertex.logProb(Nd4jIntegerTensor.scalar(upperBound)), Double.NEGATIVE_INFINITY, 1e-6);
+    }
+
+    @Test
+    public void logProbLowerBoundIsNotNegativeInfinity() {
+        UniformIntVertex testUniformVertex = new UniformIntVertex(new int[]{1, N}, lowerBound, upperBound);
+        assertNotEquals(testUniformVertex.logProb(Nd4jIntegerTensor.scalar(lowerBound)), Double.NEGATIVE_INFINITY, 1e-6);
+    }
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/UniformIntVertexTest.java]: Good new tests.

**Comment by GeorgeNash** [comment]
LGTM

**Pull request merged**